### ISSUE TITLE
GuiUnit is compiled against wrong mscorlib for XM

### ISSUE THREAD

**Issue by akoeplinger** [xamarin/xamarin-macios#5760]
### Steps to Reproduce

1. Run `make ../external/guiunit/bin/net_4_5/GuiUnit.exe XBUILD_VERBOSITY=/v:diag` in tests/
2. Observe that wrong mscorlib.dll is used: `/Library/Frameworks/Mono.framework/Versions/6.0.0/lib/mono/4.5/mscorlib.dll`

This causes an issue in mono-2019-02 since types moved from System.dll to mscorlib.dll and due to the mismatch the compiler complains about types existing in both assemblies.

Note that this is not a regression, it happens in master too but probably wasn't noticed yet since there were no conflicting types.

### Expected Behavior

The mscorlib.dll from -api is used.

### Actual Behavior

The mscorlib from `/Library/Frameworks/Mono.framework/Versions/6.0.0/lib/mono/4.5/mscorlib.dll` is used.

### Build Logs

https://gist.github.com/akoeplinger/2a617a289a7bd8db6c611d9be0a35d23

**Comment by marek-safar** [comment]
This is not a new issue but it is blocking 2019-02 integration.

@radical could you please help them to find the culprint of xammac wrong mscorlib reference?

**Comment by akoeplinger** [comment]
Note: if you add an explicit `<Reference Include="mscorlib" />` to the .csproj then it resolves to the correct assembly.

**Comment by spouliot** [comment]
AFAIK the misnamed `net_4_5` profile is used for XM **Full** so the assemblies (from mono `2019-02` and XM) should be nearly identical.

c.c. @chamons

**Comment by chamons** [comment]
```
./mcs/build/profiles/xammac.make => Modern 
./mcs/build/profiles/xammac_net_4_5.make => Full
```

**Comment by spouliot** [comment]
ok, so I don't see why mono 5.20 and `2019-02` (XM) mscorlib for Full should have a difference that prevents GuiUnit from compiling. If types were moved it should be consistent between them - and the build should succeed. 

Now if the wrong (path'ed) mscorlib was used at **runtime** that would be a bug - but, for compilation, I don't think it's an SDK issue

**Comment by marek-safar** [comment]
@spouliot the problem is that it's mixing reference and implementation assemblies (using each of them separate works),this is not a new issue, but as the API surface diverges we hit it now.

**Comment by spouliot** [comment]
hmm... that might break a lot of other, existing projects :(

**Comment by marek-safar** [comment]
Why? Because they use API which should not be available?

**Comment by spouliot** [comment]
That might be an issue too. I was thinking of existing projects with similar .csproj files, if `GuiUnit` fails then other will hit the same issue - which influence how this can (or should) be fixed.

**Comment by radical** [comment]
Could somebody also share the diag log for building the same project with 5.22.0? Meanwhile, I'm trying to build XM locally.

**Comment by radical** [comment]
Like @akoeplinger said, this is not a regression and we have the same behavior with 5.2x too.
@chamons Maybe this should be fixed up in the XM targets?

**Comment by marek-safar** [comment]
@spouliot yes, they will fail in same way but the right fix is not to mix different version of .NET framework assemblies which is what XM has been doing

**Comment by rolfbjarne** [comment]
[Diagnostic log with 5.20](https://gist.github.com/rolfbjarne/394a5dfe62bed52c56054dcd1e59969a) - builds fine
[Diagnostic log with 5.22](https://gist.github.com/rolfbjarne/0e9d9b224d890da6b569325427340b78) - does not build

**Comment by radical** [comment]
@rolfbjarne Thanks for the logs.

What is happening here is:

- `mscorlib.dll` is being used from `mono/4.5` and the other system assemblies are from `mono/4.5-api`
- GuiNet* project is built with `xibuild`

- `xibuild` [sets](https://github.com/xamarin/xamarin-macios/blob/master/tools/xibuild/Main.cs#L209) `SetToolsetProperty ("TargetFrameworkRootPath", FrameworksDirectory + Path.DirectorySeparatorChar);` which points to `/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/xbuild-frameworks`.
  - This causes `$(FrameworkPathOverride)` to be [set](https://github.com/mono/msbuild/blob/xplat-master/src/Tasks/Microsoft.Common.CurrentVersion.targets#L79) to `/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/xbuild-frameworks/.NETFramework/v4.5`, but that doesn't have a `mscorlib.dll`, so it gets [reset](https://github.com/mono/msbuild/blob/xplat-master/src/Tasks/Microsoft.Common.CurrentVersion.targets#L84) to `/Library/Frameworks/Mono.framework/Versions/5.22.0/lib/mono/4.5/`.

  - If we don't set `TargetFrameworkRoothPath`, then we get `FrameworkPathOverride = /Library/Frameworks/Mono.framework/Versions/5.22.0/lib/mono/4.5-api`, causing ` _ExplicitReference=/Library/Frameworks/Mono.framework/Versions/5.22.0/lib/mono/4.5-api/mscorlib.dll` to be used.

@chamons Am I correct that the `GuiUnit_NET_4_5` project is a `TargetFrameworkName=System` project? In the logs, I don't see that property set at all.
IIUC, this shouldn't be affecting regular user projects, built with msbuild directly. Please correct me if I'm wrong and suggest how I can reproduce it.

Also, what all tests (make targets) should I be running for this?

**Comment by radical** [comment]
@rolfbjarne @chamons Any thoughts on my last comment?

**Comment by chamons** [comment]
Sorry for missing this @radical 

The name `GuiUnit_NET_4_5` is unfortunately hard to parse, since it could be one of a number of things(XM Full, XM system, or a straight system library) based on the name alone.

I looked at the library, and it's not XM specific at all. It doesn't include any XM targets, so it would be a system library I believe.

**Comment by akoeplinger** [comment]
FYI this happens with generator projects too, I just had to do the workaround on 2019-02: https://github.com/xamarin/xamarin-macios/commit/9bd927bb7fe3eb1b14721745c500ff8b4aecd109

**Comment by radical** [comment]
@chamons np!
Ok, so the problem is when building with `xibuild`. I think then fix I suggested in https://github.com/xamarin/xamarin-macios/issues/5760#issuecomment-472457202 should work fine. With that the workaround should not be required.

**Comment by radical** [comment]
I'll open a PR for the fix.
